### DISCUSSION TITLE
Organizing justpy project folder structure and importing modules and invoking functions in jp

### DISCUSSION THREAD

**Discussion by munna123321** [elimintz/justpy#303] [Q&A]
I wanted to do ML project. Justpy happens to be good but how to organize the files into standard web application project and import modules and functions into jp?
if the code is very large how to organize justpy code into standard web application folder structure type Django/flask?
Like frontend/subfolders, templates, assets server/subfolders , authentication etc..
I did not come across importing modules into jp. How to import modules into jp?
Or can I  add packages and call functions from jp?

**Comment by elimintz** [comment]
A Justpy project does not require any special structure.
What I do is the following: I have the main file in which the import justpy and jp.justpy commands are as well as the request handlers and event handlers. Then I have some utility and specific function files from which I import to the main file as required.

**Reply by munna123321** [reply]
I want to develop ML based multipage web application.
Maintaining in a single file looks very disorganized. Need not be as large as Django type folder structure.
But should have small number of files along with assets and static files with delineation between server and frontend code.

**Comment by munna123321** [comment]
Can you give small piece of code small function file import into the main file and invoking the function.
Thanks

**Comment by elimintz** [comment]
Does this help;
https://stackoverflow.com/questions/20309456/call-a-function-from-another-file

I may be misunderstanding what the issue is

**Reply by munna123321** [reply]
Thanks , Yes let me state clearly what my requirement is. 
I want to split the project into multiple files.
main.py, client/page1.py, client/page2.py, server/server1.py, server/server2.py, server/server3.py , static/*.img , template/*.html etc..
main.py handles '/' index.html , page1.py displays data in a editor box, processed in server1.py which is fetched from database
page2.py plays video in video player processed by server2.py fetched from database or through websocket
server3.py handles rest api whose data is processed by server2.py file and then passed onto client side file.

My doubts are as follows
1)justpy is imported in multiple files , since it is a global module . Is there multiple inclusion error occurs similar to c programing ?
2) What is the visibility/scoping  of jp data across multiple files. 
ex: server1.py processes data from database and pass onto page1.py text box . Can i do it or I have to declare any global variables?
let us say html_string is formatted string. Can i pass jp.html_string from server1.py to page1.py text box?
3) server2.py  will handle API request  or video streaming request and then pass data to page2.py video player.
4) Or server3.py passes some audio/image data to client file.

My main concern is how can i pass the data across multiple files , visibility and scoping

**Reply by munna123321** [reply]
I am thinking following lines
main.py {
                  import just.py as jp
                 //do i have to import page1.py as pg1 & page2.py as pg2?? in main.py
             .   //declare menu items for page1 & page2 
                 //on_click navigates to page1 & page2
              }
page1.py {
            //display default data from database processed in server1.py
           //do i have to import server1 as ser1 
           //user selects some data in page1. server1 receives data from page1 and passes data to page1 text box
            }
page2.py {
         //user selects video data 
         //server2.py gets from database or issues api request and passes data to page2 video player
}

**Reply by munna123321** [reply]
/user selects some data in page1. server1 receives data from page1 and processes using some algorithm and passes data to page1 text box

**Reply by munna123321** [reply]
Another problem is
server2.py may have jp.html_string object formatted to display page2 html
server1.py also has jp.html_string to display page1 html.
Since jp is a global module , how is that html_string is selected?
Is it i have to declare local/global  variables of type html_string objects and assign to jp.html_string object at the of navigating?

**Reply by elimintz** [reply]
How about the following structure for main.py:
```python
import justpy as jp

from server1 import first_page_creation_function  # function returns WebPage instance
from server2 import second_page_creation_function

class GlobalData:

    data: {}

# Set the global data to what you need

@jp.SetRoute('/firstpage')
def serve_first_page(request):
    return first_page_creation_function(request, GlobalData, jp)

@jp.SetRoute('/secondpage')
def serve_second_page(request):
    return second_page_creation_function(request, GlobalData, jp)

jp.justpy()

```

**Reply by munna123321** [reply]
Thanks for suggestion. Yes this might work and will try. Thanks once again for spending your valuetime.
